dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/user/DefaultUserService.java

```diff
@@ -610,26 +617,59 @@ public boolean isAccountExpired(User user) {
 
   @Override
   @Transactional(readOnly = true)
-  public List<ErrorReport> validateUserCreateOrUpdate(User user, User currentUser) {
+  public List<ErrorReport> validateUserCreateOrUpdateAccess(User user, User currentUser) {
 
     List<ErrorReport> errors = new ArrayList<>();
 
-    if (currentUser == null || user == null) {
+    if (currentUser == null || user == null || currentUser.isSuper()) {
       return errors;
     }
 
     User userToChange = userStore.get(user.getId());
-    if (!currentUser.isSuper() && userToChange != null && userToChange.isSuper()) {
+    if (userToChange != null && userToChange.isSuper()) {
       errors.add(new ErrorReport(User.class, ErrorCode.E3041, currentUser.getUsername()));
     }
 
-    validateUserRoles(user, currentUser, errors);
-    validateUserGroups(user, currentUser, errors);
+    checkHasAccessToUserRoles(user, currentUser, errors);
+    checkHasAccessToUserGroups(user, currentUser, errors);
+
+    checkIsInOrgUnitHierarchy(user.getOrganisationUnits(), currentUser, errors);
+    checkIsInOrgUnitHierarchy(user.getDataViewOrganisationUnits(), currentUser, errors);
+    checkIsInOrgUnitHierarchy(user.getTeiSearchOrganisationUnits(), currentUser, errors);
 
     return errors;
   }
 
-  private void validateUserGroups(User user, User currentUser, List<ErrorReport> errors) {
+  private void checkIsInOrgUnitHierarchy(
+      Set<OrganisationUnit> organisationUnits, User currentUser, List<ErrorReport> errors) {
+    for (OrganisationUnit orgUnit : organisationUnits) {
+      boolean inUserHierarchy = organisationUnitService.isInUserHierarchy(currentUser, orgUnit);
+      if (!inUserHierarchy) {
+        errors.add(
+            new ErrorReport(
+                OrganisationUnit.class,
+                ErrorCode.E7617,
+                orgUnit.getUid(),
+                currentUser.getUsername()));
+      }
+    }
+  }
+
+  //  private void checkIsInOrgUnitHierarchy(
+  //      Set<OrganisationUnit> organisationUnits, User currentUser, List<ErrorReport> errors) {
+  //    for (OrganisationUnit orgUnit : organisationUnits) {
+  //      if (!orgUnit.isDescendant(currentUser.getOrganisationUnits())) {
+  //        errors.add(
+  //            new ErrorReport(
+  //                OrganisationUnit.class,
+  //                ErrorCode.E7617,
+  //                orgUnit.getUid(),
+  //                currentUser.getUsername()));
+  //      }
+  //    }
+  //  }
```

Comment:
?

Reply:
Seems that E2E tests fails when I use the orgUnit.isDescendant() method.

Reply:
I was commenting because of the commented out code, I guess it can be removed

Reply:
It is just there while I experiment with both methods